### PR TITLE
EMSUSD-101 Override primWriter

### PR DESCRIPTION
**Pull request by samuelliu-adsk**
- Add `CanExport` function to primWriter for overriding default primWriter
- Overload the Register function for backward support
- Add `Poke `function for cases when only wants to check for plugin register but not really find a primWriter

### CODE REVIEW

**Review by calligrapheric — APPROVED** [review]

**Review by AramAzhari-adsk — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]
Same comment as with the Reader PR. I'm not familiar with this code so I'll let the other code reviewers provided the feedback.

File: lib/mayaUsd/fileio/primWriterRegistry.cpp [hunk 1]
```diff
@@ -42,9 +42,75 @@ TF_DEFINE_PRIVATE_TOKENS(
...
-static std::set<std::string> _mayaTypesThatDoNotCreatePrims;
+namespace {
+struct _RegistryEntry
+{
+    UsdMayaPrimWriterRegistry::ContextPredicateFn _pred;
+    UsdMayaPrimWriterRegistry::WriterFactoryFn    _writer;
+    int                                           _index;
+};
+
+// typedef std::map<std::string, UsdMayaPrimWriterRegistry::WriterFactoryFn> _Registry;
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primWriterRegistry.cpp:53]: I think this typedef in comment was left-over and probably should be removed.

File: lib/mayaUsd/fileio/primWriterRegistry.h [hunk 2]
```diff
@@ -109,7 +142,14 @@ struct UsdMayaPrimWriterRegistry
...
     MAYAUSD_CORE_PUBLIC
-    static WriterFactoryFn Find(const std::string& mayaTypeName);
+    static WriterFactoryFn Find(
+        const std::string&          mayaTypeName,
+        const UsdMayaJobExportArgs& exportArgs,
+        const MObject&              exportObj);
+
+    /// \brief Check for external primWriter for \p mayaTypeName.
+    MAYAUSD_CORE_PUBLIC
+    static void Poke(const std::string& mayaTypeName);
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primWriterRegistry.h:153]: IDK if a better description and function name could be found. I read both the docs and function name and could not figure what the function did. The fact it returns nothing is adding to the confusion as it is unclear what the result of calling this is.  Since it seems to trigger scanning plugins for support for that Maya type, the description and function name could hint at that.

File: lib/mayaUsd/fileio/primWriterRegistry.cpp [hunk 3]
```diff
@@ -74,32 +157,57 @@ void UsdMayaPrimWriterRegistry::RegisterRaw(
...
+UsdMayaPrimWriterRegistry::WriterFactoryFn UsdMayaPrimWriterRegistry::Find(
+    const std::string&          mayaTypeName,
+    const UsdMayaJobExportArgs& exportArgs,
+    const MObject&              exportObj)
 {
     TfRegistryManager::GetInstance().SubscribeTo<UsdMayaPrimWriterRegistry>();
 
-    // unfortunately, usdTypeName is diff from the tfTypeName which we use to
-    // register.  do the conversion here.
-    WriterFactoryFn ret = nullptr;
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primWriterRegistry.cpp:84]: IDK what that erased comment about usdTypeName vs tfTypeName was about... I guess it is now irrelevant?

File: lib/mayaUsd/fileio/primWriterRegistry.cpp [hunk 4]
```diff
@@ -74,32 +157,57 @@ void UsdMayaPrimWriterRegistry::RegisterRaw(
...
     UsdMaya_RegistryHelper::FindAndLoadMayaPlug(SCOPE, mayaTypeName);
 
+    it = _Find(mayaTypeName, exportArgs, exportObj);
+
+    if (it != _reg.end()) {
+        return it->second._writer;
+    }
+
     // ideally something just registered itself.  if not, we at least put it in
     // the registry in case we encounter it again.
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/fileio/primWriterRegistry.cpp:183]: This comment should be moved a few lines up before the 2nd _Find.

File: lib/mayaUsd/python/wrapPrimWriter.cpp [hunk 5]
```diff
@@ -206,20 +206,36 @@ class PrimWriterWrapper
...
         {
             size_t classIndex = RegisterPythonObject(cl, GetKey(cl, mayaTypeName));
-            if (classIndex != UsdMayaPythonObjectRegistry::UPDATED) {
-                // Return a new factory function:
-                return FactoryFnWrapper { classIndex };
-            } else {
-                // We already registered a factory function for this purpose:
-                return nullptr;
-            }
+            updated = classIndex == UsdMayaPythonObjectRegistry::UPDATED;
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/python/wrapPrimWriter.cpp:236]: I'm never sure about = vs == priority, so as a dev reading the code, parenthesis would make me feel safer.

File: lib/mayaUsd/python/wrapPrimWriter.cpp [hunk 6]
```diff
@@ -246,10 +262,10 @@ class PrimWriterWrapper
...
     {
-        UsdMayaPrimWriterRegistry::WriterFactoryFn fn
-            = FactoryFnWrapper::Register(cl, mayaTypeName);
-        if (fn) {
-            UsdMayaPrimWriterRegistry::Register(mayaTypeName, fn, true);
+        bool             updated = false;
+        FactoryFnWrapper fn = FactoryFnWrapper::Register(cl, mayaTypeName, updated);
+        if (!updated) {
+            UsdMayaPrimWriterRegistry::Register(mayaTypeName, fn, fn, true);
         }
```
Comment (pierrebai-adsk) [inline lib/mayaUsd/python/wrapPrimWriter.cpp:269]: I guess the fact the class has an operator() allows it to be both the predicate and factory function, but a comment explaining why fn gets passed twice would make the code clearer

File: test/lib/mayaUsd/fileio/testPrimWriter.py [hunk 7]
```diff
@@ -49,6 +49,10 @@ def __init__(self, *args, **kwargs):
         self._SetUsdPrim(usdPrim)
         primWriterTest.InitCalled = True
 
+    @classmethod
+    def CanExport(cls, exportArgs, exportObj=None):
+        return mayaUsdLib.PrimWriter.ContextSupport.Supported
```
Comment (pierrebai-adsk) [inline test/lib/mayaUsd/fileio/testPrimWriter.py:54]: I think adding a flag "CanExportCalled" like in the other test would help verify the functionality.

**Review by pierrebai-adsk — CHANGES_REQUESTED** [review]
Just a few questions and suggestions, otherwise looks great.

File: lib/mayaUsd/fileio/primWriterRegistry.cpp (hunk 4 above)
Comment (samuelliu-adsk) [reply lib/mayaUsd/fileio/primWriterRegistry.cpp:183]: This comment was trying to explain what the code below does so I'll reword it to make it clearer

**Review by pierrebai-adsk — APPROVED** [review]
